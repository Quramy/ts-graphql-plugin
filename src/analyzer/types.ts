import { TsGraphQLPluginConfigOptions } from '../types';
import { TypeGenAddonFactory } from '../typegen';

export type OperationType = 'query' | 'mutation' | 'subscription' | 'fragment' | 'complex' | 'other';

export type TsGraphQLPluginConfig = Omit<TsGraphQLPluginConfigOptions, 'typegen'> & {
  typegen: {
    addonFactories: TypeGenAddonFactory[];
  };
};

export interface ManifestDocumentEntry {
  fileName: string;
  type: OperationType;
  operationName?: string;
  fragmentName?: string;
  body: string;
  tag?: string;
  documentStart: { line: number; character: number };
  documentEnd: { line: number; character: number };
  templateLiteralNodeStart: { line: number; character: number };
  templateLiteralNodeEnd: { line: number; character: number };
}

export interface ManifestOutput {
  documents: ManifestDocumentEntry[];
}
