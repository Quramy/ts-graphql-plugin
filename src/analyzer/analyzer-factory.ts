import path from 'path';
import ts from 'typescript';
import { Analyzer } from './analyzer';
import { TsGraphQLPluginConfigOptions } from '../types';
import { ScriptHost } from '../ts-ast-util';
import { SchemaManagerFactory, createSchemaManagerHostFromTSGqlPluginConfig } from '../schema-manager';
import { ErrorWithoutLocation } from '../errors';
import { TypeGenAddonFactory } from '../typegen';

const NO_PLUGCN_SETTING_ERROR_MESSAGE = `tsconfig.json should have ts-graphql-plugin setting. Add the following:
  "compilerOptions": {
    "plugins": [
      {
        "name": "ts-graphql-plugin",
        "schema": "shema.graphql",   /* Path to your GraphQL schema */
        "tag": "gql"                 /* Template tag function name */
      }
    ]
  }`;

class TypegenAddonLoadError extends ErrorWithoutLocation {
  constructor(addonName: string) {
    const message = `Fail to load typegen add-on. Confirm "${addonName}" points correct add-on script.`;
    super(message);
  }
}

function loadAddonFactories(pluginConfig: TsGraphQLPluginConfigOptions, prjRootPath: string) {
  if (!pluginConfig.typegen?.addons || !Array.isArray(pluginConfig.typegen?.addons)) return [];
  const factories = pluginConfig.typegen.addons.map(addonName => {
    const addonPath =
      !path.isAbsolute(addonName) && addonName.startsWith('.')
        ? path.resolve(prjRootPath, addonName)
        : path.normalize(addonName);
    try {
      require.resolve(addonPath);
    } catch {
      throw new TypegenAddonLoadError(addonName);
    }
    return require(addonPath) as TypeGenAddonFactory;
  });
  return [...new Set(factories)];
}

export class AnalyzerFactory {
  createAnalyzerAndScriptHostFromProjectPath(
    projectPath: string,
    debug: (msg: string) => void = () => {},
    currentDirectory = process.cwd(),
  ) {
    const { pluginConfigOptions, tsconfig, prjRootPath } = this._readTsconfig(projectPath);
    const scriptHost = new ScriptHost(currentDirectory, tsconfig.options);
    tsconfig.fileNames.forEach(fileName => scriptHost.readFile(fileName));
    const addonFactories = loadAddonFactories(pluginConfigOptions, prjRootPath);
    const pluginConfig = {
      ...pluginConfigOptions,
      typegen: {
        addonFactories,
      },
    };
    const schemaManagerHost = createSchemaManagerHostFromTSGqlPluginConfig(pluginConfig, prjRootPath, debug);
    const schemaManager = new SchemaManagerFactory(schemaManagerHost).create();
    const analyzer = new Analyzer(pluginConfig, prjRootPath, scriptHost, schemaManager, debug);
    return { analyzer, scriptHost };
  }

  createAnalyzerFromProjectPath(
    projectPath: string,
    debug: (msg: string) => void = () => {},
    currentDirectory = process.cwd(),
  ) {
    return this.createAnalyzerAndScriptHostFromProjectPath(projectPath, debug, currentDirectory).analyzer;
  }

  private _readTsconfig(project: string) {
    const currentDirectory = ts.sys.getCurrentDirectory();
    const ppath = path.isAbsolute(project) ? path.resolve(currentDirectory, project) : project;
    let configPath: string | undefined = undefined;
    if (ts.sys.fileExists(ppath)) {
      configPath = ppath;
    } else if (ts.sys.directoryExists(ppath) && ts.sys.fileExists(path.join(ppath, 'tsconfig.json'))) {
      configPath = path.join(ppath, 'tsconfig.json');
    }
    if (!configPath) {
      throw new Error(`tsconfig file not found: ${project}`);
    }
    const tsconfig = ts.getParsedCommandLineOfConfigFile(configPath, {}, ts.sys as any);
    if (!tsconfig) {
      throw new Error(`Failed to parse: ${configPath}`);
    }
    const prjRootPath = path.dirname(configPath);
    const plugins = tsconfig.options.plugins;
    if (!plugins || !Array.isArray(plugins)) {
      throw new Error(NO_PLUGCN_SETTING_ERROR_MESSAGE);
    }
    const found = (plugins as any[]).find((p: any) => (p._name || p.name) === 'ts-graphql-plugin');
    if (!found) {
      throw new Error(NO_PLUGCN_SETTING_ERROR_MESSAGE);
    }
    const pluginConfigOptions = found as TsGraphQLPluginConfigOptions;
    return {
      tsconfig,
      pluginConfigOptions,
      prjRootPath,
    };
  }
}
