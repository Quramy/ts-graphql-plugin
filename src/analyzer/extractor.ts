import ts from 'typescript';
import { parse, print, DocumentNode, GraphQLError } from 'graphql';
import { visit } from 'graphql/language';
import { isTagged, ScriptSourceHelper, ResolvedTemplateInfo } from '../ts-ast-util';
import { ManifestOutput, ManifestDocumentEntry, OperationType } from './types';
import { ErrorWithLocation, ERROR_CODES } from '../errors';
import { detectDuplicatedFragments } from '../gql-ast-util';

export type ExtractorOptions = {
  removeDuplicatedFragments: boolean;
  scriptSourceHelper: ScriptSourceHelper;
  debug: (msg: string) => void;
};

export type ExtractTemplateResolveErrorResult = {
  fileName: string;
  templateNode: ts.NoSubstitutionTemplateLiteral | ts.TemplateExpression;
  resolveTemplateError: {
    message: string;
    start: number;
    end: number;
  };
  documentNode: undefined;
  resolevedTemplateInfo: undefined;
  graphqlError: undefined;
};

export type ExtractGraphQLErrorResult = {
  fileName: string;
  templateNode: ts.NoSubstitutionTemplateLiteral | ts.TemplateExpression;
  resolevedTemplateInfo: ResolvedTemplateInfo;
  resolveTemplateError: undefined;
  graphqlError: GraphQLError;
  resolveTemplateErrorMessage: undefined;
  documentNode: undefined;
};

export type ExtractSucceededResult = {
  fileName: string;
  templateNode: ts.NoSubstitutionTemplateLiteral | ts.TemplateExpression;
  documentNode: DocumentNode;
  resolevedTemplateInfo: ResolvedTemplateInfo;
  resolveTemplateError: undefined;
  graphqlError: undefined;
  resolveTemplateErrorMessage: undefined;
};

export type ExtractResult = ExtractTemplateResolveErrorResult | ExtractGraphQLErrorResult | ExtractSucceededResult;

export class Extractor {
  private readonly _removeDuplicatedFragments: boolean;
  private readonly _helper: ScriptSourceHelper;
  private readonly _debug: (msg: string) => void;

  constructor({ debug, removeDuplicatedFragments, scriptSourceHelper }: ExtractorOptions) {
    this._removeDuplicatedFragments = removeDuplicatedFragments;
    this._helper = scriptSourceHelper;
    this._debug = debug;
  }

  extract(files: string[], tagName?: string): ExtractResult[] {
    const results: ExtractResult[] = [];
    this._debug('Extract template literals from: ');
    this._debug(files.map(f => ' ' + f).join(',\n'));
    files.forEach(fileName => {
      const nodes = this._helper
        .getAllNodes(fileName, node => ts.isTemplateExpression(node) || ts.isNoSubstitutionTemplateLiteral(node))
        .filter(node => (tagName ? isTagged(node, tagName) : true)) as (
        | ts.TemplateExpression
        | ts.NoSubstitutionTemplateLiteral
      )[];
      nodes.forEach(node => {
        const { resolvedInfo, resolveErrors } = this._helper.resolveTemplateLiteral(fileName, node);
        if (!resolvedInfo) {
          resolveErrors
            .filter(re => re.fileName === fileName)
            .forEach(resolveError => {
              results.push({
                fileName,
                templateNode: node,
                resolveTemplateError: {
                  message: ERROR_CODES.templateIsTooComplex.message,
                  start: resolveError.start,
                  end: resolveError.end,
                },
                resolevedTemplateInfo: undefined,
                graphqlError: undefined,
                documentNode: undefined,
              });
            });
        } else {
          results.push({
            fileName,
            templateNode: node,
            resolevedTemplateInfo: resolvedInfo,
          } as any);
        }
      });
    });
    return results.map(result => {
      if (!result.resolevedTemplateInfo) return result;
      try {
        const rawDocumentNode = parse(result.resolevedTemplateInfo.combinedText);
        if (!this._removeDuplicatedFragments) {
          return {
            ...result,
            documentNode: rawDocumentNode,
          } as ExtractSucceededResult;
        }
        const duplicatedInfo = detectDuplicatedFragments(rawDocumentNode);
        const updatedResolvedInfo = duplicatedInfo.reduce(
          (acc, fragmentInfo) => this._helper.updateTemplateLiteralInfo(acc, fragmentInfo),
          result.resolevedTemplateInfo,
        );
        const documentNode = parse(updatedResolvedInfo.combinedText);
        return {
          ...result,
          documentNode,
          resolevedTemplateInfo: updatedResolvedInfo,
        } as ExtractSucceededResult;
      } catch (error) {
        if (error instanceof GraphQLError) {
          return {
            ...result,
            graphqlError: error,
          } as ExtractGraphQLErrorResult;
        } else {
          throw error;
        }
      }
    });
  }

  pickupErrors(
    extractResults: ExtractResult[],
    { ignoreGraphQLError }: { ignoreGraphQLError: boolean } = { ignoreGraphQLError: false },
  ) {
    const errors: ErrorWithLocation[] = [];
    extractResults.forEach(r => {
      if (r.resolveTemplateError) {
        errors.push(
          new ErrorWithLocation(r.resolveTemplateError.message, {
            fileName: r.fileName,
            severity: 'Warn',
            content: r.templateNode.getSourceFile().getFullText(),
            start: r.resolveTemplateError.start,
            end: r.resolveTemplateError.end,
          }),
        );
      } else if (!ignoreGraphQLError && r.graphqlError) {
        const innerLoc = r.graphqlError.locations && r.graphqlError.locations[0];
        if (!innerLoc) {
          errors.push(
            new ErrorWithLocation(r.graphqlError.message, {
              fileName: r.fileName,
              content: r.templateNode.getSourceFile().getFullText(),
              start: r.templateNode.getStart(),
              end: r.templateNode.getEnd(),
            }),
          );
        } else if (r.resolevedTemplateInfo) {
          const info = r.resolevedTemplateInfo;
          const innerPosition = info.convertInnerLocation2InnerPosition({
            line: innerLoc.line - 1,
            character: innerLoc.column - 1,
          });
          const start = info.getSourcePosition(innerPosition).pos;
          errors.push(
            new ErrorWithLocation(r.graphqlError.message, {
              fileName: r.fileName,
              content: r.templateNode.getSourceFile().getFullText(),
              start,
              end: start + 1,
            }),
          );
        }
      }
    });
    return errors;
  }

  getDominantDefinition(result: ExtractSucceededResult) {
    let type: OperationType | undefined;
    const definedFragmentNames: string[] = [];
    const referencedFragmentNames: string[] = [];
    let operationName: string | undefined;
    visit(result.documentNode, {
      FragmentDefinition(node) {
        if (!type) {
          type = 'fragment';
        }
        definedFragmentNames.push(node.name.value);
      },
      FragmentSpread(node) {
        referencedFragmentNames.push(node.name.value);
      },
      OperationDefinition(node) {
        if (!type || type === 'fragment') {
          type = node.operation;
        } else {
          type = 'complex';
        }
        if (!operationName) {
          operationName = node.name ? node.name.value : 'ANONYMOUS_QUERY';
        } else {
          operationName = 'MULTIPLE_OPERATIONS';
        }
      },
    });
    const noReferedFragmentNames = definedFragmentNames.filter(defName =>
      referencedFragmentNames.every(n => defName !== n),
    );
    return { type, operationName, fragmentName: noReferedFragmentNames[noReferedFragmentNames.length - 1] };
  }

  toManifest(extractResults: ExtractResult[], tagName: string = ''): ManifestOutput {
    const documents = extractResults
      .filter(r => !!r.documentNode)
      .map(result => {
        const r = result as ExtractSucceededResult;
        const { type, operationName, fragmentName } = this.getDominantDefinition(r);
        return {
          fileName: r.fileName,
          type: type || 'other',
          operationName,
          fragmentName,
          body: print(r.documentNode!),
          tag: tagName,
          templateLiteralNodeStart: this._helper.getLineAndChar(r.fileName, r.templateNode.getStart()),
          templateLiteralNodeEnd: this._helper.getLineAndChar(r.fileName, r.templateNode.getEnd()),
          documentStart: this._helper.getLineAndChar(r.fileName, r.templateNode.getStart() + 1),
          documentEnd: this._helper.getLineAndChar(r.fileName, r.templateNode.getEnd() - 1),
        } as ManifestDocumentEntry;
      });
    return {
      documents,
    };
  }
}
