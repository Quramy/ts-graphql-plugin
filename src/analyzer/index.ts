export * from './types';
export { Analyzer } from './analyzer';
export { AnalyzerFactory } from './analyzer-factory';
export type {
  ExtractResult,
  ExtractSucceededResult,
  ExtractGraphQLErrorResult,
  ExtractTemplateResolveErrorResult,
} from './extractor';
