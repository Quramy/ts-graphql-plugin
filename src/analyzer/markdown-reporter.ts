import path from 'path';
import { ManifestOutput, ManifestDocumentEntry } from './types';

export type ToMarkdownContentOptions = {
  baseDir: string;
  outputDir: string;
  ignoreFragments?: boolean;
};

function createLinkPath(fileName: string, baseDir: string, outputDir: string) {
  const displayPath = path.isAbsolute(fileName) ? path.relative(baseDir, fileName) : fileName;
  const linkPath = path.isAbsolute(fileName) ? path.relative(outputDir, fileName) : fileName;
  return { displayPath, linkPath };
}

function createSection(
  sectionName: string,
  operationDocs: ManifestDocumentEntry[],
  baseDir: string,
  outputDir: string,
) {
  const h2 = '## ' + sectionName[0].toUpperCase() + sectionName.slice(1);
  return (
    h2 +
    '\n' +
    operationDocs
      .map(doc => {
        const { displayPath, linkPath } = createLinkPath(doc.fileName, baseDir, outputDir);
        return `
### ${doc.type !== 'fragment' ? doc.operationName || 'anonymous' : doc.fragmentName}

\`\`\`graphql
${doc.body.trim()}
\`\`\`

From [${displayPath}:${doc.documentStart.line + 1}:${doc.documentStart.character + 1}](${linkPath}#L${
          doc.documentStart.line + 1
        }-L${doc.documentEnd.line + 1})
    `;
      })
      .join('\n')
  );
}

export class MarkdownReporter {
  toMarkdownConntent(
    manifest: ManifestOutput,
    { ignoreFragments = true, baseDir, outputDir }: ToMarkdownContentOptions,
  ) {
    if (!manifest.documents.length) return null;
    const outs = ['# Extracted GraphQL Operations'];
    const groupedDocs = {
      queries: [],
      mutations: [],
      subscriptions: [],
      fragments: [],
    } as { [key: string]: ManifestDocumentEntry[] };
    manifest.documents.forEach(doc => {
      switch (doc.type) {
        case 'query':
          groupedDocs.queries.push(doc);
          break;
        case 'mutation':
          groupedDocs.mutations.push(doc);
          break;
        case 'subscription':
          groupedDocs.subscriptions.push(doc);
          break;
        case 'fragment':
          if (!ignoreFragments) groupedDocs.fragments.push(doc);
          break;
        default:
          break;
      }
    });
    Object.entries(groupedDocs).forEach(([name, docs]) => {
      if (docs.length) {
        outs.push(createSection(name, docs, baseDir, outputDir));
      }
    });
    outs.push('---');
    outs.push('Extracted by [ts-graphql-plugin](https://github.com/Quramy/ts-graphql-plugin)');
    return outs.join('\n');
  }
}
