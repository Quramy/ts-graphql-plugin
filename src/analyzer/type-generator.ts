import path from 'path';
import ts from 'typescript';
import { GraphQLSchema } from 'graphql/type';

import { TsGqlError, ErrorWithLocation } from '../errors';
import { mergeAddons, TypeGenVisitor, TypeGenError, TypeGenAddonFactory, TypeGenVisitorAddonContext } from '../typegen';
import { dasherize } from '../string-util';
import { OutputSource, createOutputSource } from '../ts-ast-util';
import { Extractor, ExtractSucceededResult } from './extractor';

export type TypeGeneratorOptions = {
  prjRootPath: string;
  extractor: Extractor;
  debug: (msg: string) => void;
  tag: string | undefined;
  addonFactories: TypeGenAddonFactory[];
};

export class TypeGenerator {
  private readonly _prjRootPath: string;
  private readonly _extractor: Extractor;
  private readonly _tag: string | undefined;
  private readonly _addonFactories: TypeGenAddonFactory[];
  private readonly _debug: (msg: string) => void;
  private readonly _printer: ts.Printer;

  constructor({ prjRootPath, extractor, tag, addonFactories, debug }: TypeGeneratorOptions) {
    this._prjRootPath = prjRootPath;
    this._extractor = extractor;
    this._tag = tag;
    this._addonFactories = addonFactories;
    this._debug = debug;
    this._printer = ts.createPrinter({ newLine: ts.NewLineKind.LineFeed, removeComments: false });
  }

  createAddon({
    schema,
    extractedResult,
    outputSource,
  }: {
    schema: GraphQLSchema;
    extractedResult: ExtractSucceededResult;
    outputSource: OutputSource;
  }) {
    const context: TypeGenVisitorAddonContext = {
      schema,
      source: outputSource,
      extractedInfo: {
        fileName: extractedResult.fileName,
        tsTemplateNode: extractedResult.templateNode,
        tsSourceFile: extractedResult.templateNode.getSourceFile(),
      },
    };
    const addons = this._addonFactories.map(factory => factory(context));
    return { addon: mergeAddons(addons), context };
  }

  generateTypes({ files, schema }: { files: string[]; schema: GraphQLSchema }) {
    const extractedResults = this._extractor.extract(files, this._tag);
    const extractedErrors = this._extractor.pickupErrors(extractedResults);
    if (extractedErrors.length) {
      this._debug(`Found ${extractedErrors.length} extraction errors.`);
    }
    const typegenErrors: TsGqlError[] = [];
    const visitor = new TypeGenVisitor({ schema });
    const outputSourceFiles: { fileName: string; content: string }[] = [];
    extractedResults.forEach(extractedResult => {
      if (extractedResult.documentNode) {
        const { type, fragmentName, operationName } = this._extractor.getDominantDefinition(extractedResult);
        if (type === 'complex') {
          const fileName = extractedResult.fileName;
          const content = extractedResult.templateNode.getSourceFile().getFullText();
          const start = extractedResult.templateNode.getStart();
          const end = extractedResult.templateNode.getEnd();
          const errorContent = { fileName, content, start, end };
          const error = new ErrorWithLocation('This document node has complex operations.', errorContent);
          typegenErrors.push(error);
          return;
        }
        const operationOrFragmentName = type === 'fragment' ? fragmentName : operationName;
        if (!operationOrFragmentName) return;
        const outputFileName = path.resolve(
          path.dirname(extractedResult.fileName),
          '__generated__',
          dasherize(operationOrFragmentName) + '.ts',
        );
        try {
          const outputSource = createOutputSource({ outputFileName });
          const { addon } = this.createAddon({ schema, outputSource, extractedResult });
          const outputSourceFile = visitor.visit(extractedResult.documentNode, { outputSource, addon });
          const content = this._printer.printFile(outputSourceFile);
          outputSourceFiles.push({ fileName: outputFileName, content });
          this._debug(
            `Create type source file '${path.relative(this._prjRootPath, outputFileName)}' from '${path.relative(
              this._prjRootPath,
              extractedResult.fileName,
            )}'.`,
          );
        } catch (error) {
          if (error instanceof TypeGenError) {
            const sourcePosition = extractedResult.resolevedTemplateInfo.getSourcePosition(error.node.loc!.start);
            if (sourcePosition.isInOtherExpression) return;
            const fileName = extractedResult.fileName;
            const content = extractedResult.templateNode.getSourceFile().getFullText();
            const start = sourcePosition.pos;
            const end = extractedResult.resolevedTemplateInfo.getSourcePosition(error.node.loc!.end).pos;
            const errorContent = { fileName, content, start, end };
            const translatedError = new ErrorWithLocation(error.message, errorContent);
            typegenErrors.push(translatedError);
          } else {
            throw error;
          }
        }
      }
    });
    return { errors: [...extractedErrors, ...typegenErrors], outputSourceFiles };
  }
}
