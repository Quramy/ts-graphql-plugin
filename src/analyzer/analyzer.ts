import ts from 'typescript';
import path from 'path';
import { ScriptSourceHelper } from '../ts-ast-util/types';
import { Extractor } from './extractor';
import { createScriptSourceHelper } from '../ts-ast-util';
import { SchemaManager, SchemaBuildErrorInfo } from '../schema-manager/schema-manager';
import { TsGqlError, ErrorWithLocation, ErrorWithoutLocation } from '../errors';
import { location2pos } from '../string-util';
import { validate } from './validator';
import { ManifestOutput, TsGraphQLPluginConfig } from './types';
import { MarkdownReporter } from './markdown-reporter';
import { TypeGenerator } from './type-generator';

class TsGqlConfigError extends ErrorWithoutLocation {
  constructor() {
    const message =
      'No GraphQL schema. Confirm your ts-graphql-plugin\'s "schema" configuration at tsconfig.json\'s compilerOptions.plugins section.';
    super(message);
  }
}

export function convertSchemaBuildErrorsToErrorWithLocation(errorInfo: SchemaBuildErrorInfo) {
  const fileName = errorInfo.fileName;
  const content = errorInfo.fileContent;
  if (errorInfo.locations && errorInfo.locations[0]) {
    const start = location2pos(errorInfo.fileContent, errorInfo.locations[0]);
    const end = start + 1;
    const errorContent = { fileName, content, start, end };
    return new ErrorWithLocation(errorInfo.message, errorContent);
  } else {
    const start = 0;
    const end = content.length - 1;
    const errorContent = { fileName, content, start, end };
    return new ErrorWithLocation(errorInfo.message, errorContent);
  }
}

export class Analyzer {
  private _extractor: Extractor;
  private _scriptSourceHelper: ScriptSourceHelper;
  private _typeGenerator: TypeGenerator;

  constructor(
    private readonly _pluginConfig: TsGraphQLPluginConfig,
    private readonly _prjRootPath: string,
    private readonly _languageServiceHost: ts.LanguageServiceHost,
    private readonly _schemaManager: SchemaManager,
    private readonly _debug: (msg: string) => void,
  ) {
    const langService = ts.createLanguageService(this._languageServiceHost);
    this._scriptSourceHelper = createScriptSourceHelper({
      languageService: langService,
      languageServiceHost: this._languageServiceHost,
    });
    this._extractor = new Extractor({
      removeDuplicatedFragments: this._pluginConfig.removeDuplicatedFragments === false ? false : true,
      scriptSourceHelper: this._scriptSourceHelper,
      debug: this._debug,
    });
    this._typeGenerator = new TypeGenerator({
      prjRootPath: this._prjRootPath,
      extractor: this._extractor,
      tag: this._pluginConfig.tag,
      addonFactories: this._pluginConfig.typegen.addonFactories,
      debug: this._debug,
    });
  }

  getPluginConfig() {
    return this._pluginConfig;
  }

  extract(fileNameList?: string[]) {
    const results = this._extractor.extract(
      fileNameList || this._languageServiceHost.getScriptFileNames(),
      this._pluginConfig.tag,
    );
    const errors = this._extractor.pickupErrors(results);
    return [errors, results] as const;
  }

  extractToManifest() {
    const [errors, results] = this.extract();
    const manifest = this._extractor.toManifest(results, this._pluginConfig.tag);
    return [errors, manifest] as const;
  }

  async validate() {
    const [schemaErrors, schema] = await this._getSchema();
    if (!schema) return { errors: schemaErrors };
    const [extractedErrors, results] = this.extract();
    if (extractedErrors.length) {
      this._debug(`Found ${extractedErrors.length} extraction errors.`);
    }
    return {
      errors: [...schemaErrors, ...extractedErrors, ...validate(results, schema)],
      extractedResults: results,
      schema,
    };
  }

  report(outputFileName: string, manifest?: ManifestOutput, ignoreFragments = true) {
    const reporter = new MarkdownReporter();
    const reportOptions = {
      baseDir: this._prjRootPath,
      ignoreFragments,
      outputDir: path.dirname(outputFileName),
    };
    if (manifest) {
      return [[] as TsGqlError[], reporter.toMarkdownConntent(manifest, reportOptions)] as const;
    } else {
      const [errors, extractedManifest] = this.extractToManifest();
      return [errors, reporter.toMarkdownConntent(extractedManifest, reportOptions)] as const;
    }
  }

  async typegen() {
    const [schemaErrors, schema] = await this._getSchema();
    if (!schema) return { errors: schemaErrors };
    const { errors, outputSourceFiles } = this._typeGenerator.generateTypes({
      schema,
      files: this._languageServiceHost.getScriptFileNames(),
    });
    return { errors: [...schemaErrors, ...errors], outputSourceFiles };
  }

  private async _getSchema() {
    const errors: TsGqlError[] = [];
    const { schema, errors: schemaBuildErrors } = await this._schemaManager.waitSchema();
    if (schemaBuildErrors) {
      schemaBuildErrors.forEach(info => errors.push(convertSchemaBuildErrorsToErrorWithLocation(info)));
    }
    if (!schema && !errors.length) {
      errors.push(new TsGqlConfigError());
    }
    return [errors, schema] as const;
  }
}
