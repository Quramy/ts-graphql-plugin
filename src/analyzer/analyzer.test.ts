import { Analyzer } from './analyzer';
import { createTestingLanguageServiceAndHost } from '../ts-ast-util/testing/testing-language-service';
import { createTestingSchemaManagerHost } from '../schema-manager/testing/testing-schema-manager-host';
import { SchemaManagerFactory } from '../schema-manager/schema-manager-factory';
import { TsGraphQLPluginConfig } from './types';

type CreateTestingAnalyzerOptions = {
  sdl: string;
  files?: { fileName: string; content: string }[];
  localSchemaExtension?: { fileName: string; content: string };
};

function createTestingAnalyzer({ files: sourceFiles = [], sdl, localSchemaExtension }: CreateTestingAnalyzerOptions) {
  const files = [...sourceFiles];
  files.push({ fileName: '/schema.graphql', content: sdl });
  if (localSchemaExtension) {
    files.push(localSchemaExtension);
  }
  const { languageServiceHost } = createTestingLanguageServiceAndHost({ files: sourceFiles });
  const pluginConfig: TsGraphQLPluginConfig = {
    name: 'ts-graphql-plugin',
    schema: '/schema.graphql',
    localSchemaExtensions: localSchemaExtension ? [localSchemaExtension.fileName] : [],
    removeDuplicatedFragments: true,
    tag: 'gql',
    typegen: {
      addonFactories: [],
    },
  };
  const schemaManagerHost = createTestingSchemaManagerHost({
    ...pluginConfig,
    prjRootPath: '',
    files,
  });
  return new Analyzer(
    pluginConfig,
    '',
    languageServiceHost,
    new SchemaManagerFactory(schemaManagerHost).create(),
    () => {},
  );
}

const simpleSources = {
  sdl: `
  type Query {
    hello: String!
  }
  `,
  files: [
    {
      fileName: 'main.ts',
      content: 'const query = gql`query MyQuery { hello }`;',
    },
  ],
};

const noSchemaPrj = {
  sdl: '',
  files: [
    {
      fileName: 'main.ts',
      content: 'const query = gql`query MyQuery { hello }`;',
    },
  ],
};

const extensionErrorPrj = {
  sdl: `
  type Query {
    hello: String!
  }
  `,
  files: [
    {
      fileName: 'main.ts',
      content: 'const query = gql`query MyQuery { helloo }`;',
    },
  ],
  localSchemaExtension: {
    fileName: '/extension.graphql',
    content: 'hogehoge',
  },
};

const semanticErrorPrj = {
  sdl: `
  type Query {
    hello: String!
  }
  `,
  files: [
    {
      fileName: 'main.ts',
      content: 'const query = gql`query MyQuery { helloo }`;',
    },
  ],
};

const semanticWarningPrj = {
  sdl: `
  type Query {
    hello: String! @deprecated(reason: "don't use")
  }
  `,
  files: [
    {
      fileName: 'main.ts',
      content: 'const query = gql`query MyQuery { hello }`;',
    },
  ],
};

describe(Analyzer, () => {
  describe(Analyzer.prototype.extractToManifest, () => {
    it('should extract manifest', () => {
      const analyzer = createTestingAnalyzer(simpleSources);
      expect(analyzer.extractToManifest()).toMatchSnapshot();
    });
  });

  describe(Analyzer.prototype.validate, () => {
    it('should validate project with normal project', async () => {
      const analyzer = createTestingAnalyzer(simpleSources);
      const { errors, schema } = await analyzer.validate();
      expect(errors.length).toBe(0);
      expect(schema).toBeTruthy();
    });

    it('should report error when no schema', async () => {
      const analyzer = createTestingAnalyzer(noSchemaPrj);
      const { errors } = await analyzer.validate();
      expect(errors.length).toBe(1);
      expect(errors[0].message).toMatchSnapshot();
    });

    it('should validate project with schema error project', async () => {
      const analyzer = createTestingAnalyzer(extensionErrorPrj);
      const { errors, schema } = await analyzer.validate();
      expect(errors.length).toBe(1);
      expect(errors[0].message).toMatchSnapshot();
      expect(schema).toBeFalsy();
    });

    it('should validate project with semantic error project', async () => {
      const analyzer = createTestingAnalyzer(semanticErrorPrj);
      const { errors, schema } = await analyzer.validate();
      expect(errors.length).toBe(1);
      expect(schema).toBeTruthy();
    });

    it('should validate project with semantic error project', async () => {
      const analyzer = createTestingAnalyzer(semanticWarningPrj);
      const { errors, schema } = await analyzer.validate();
      expect(errors.length).toBe(1);
      expect(schema).toBeTruthy();
    });
  });

  describe(Analyzer.prototype.report, () => {
    it('should create markdown report', () => {
      const analyzer = createTestingAnalyzer(simpleSources);
      const [errors, output] = analyzer.report('out.md');
      expect(errors.length).toBe(0);
      expect(output).toMatchSnapshot();
    });

    it('should create markdown report from manifest', () => {
      const analyzer = createTestingAnalyzer(simpleSources);
      const manifestOutput = analyzer.extractToManifest();
      const [errors, output] = analyzer.report('out.md', manifestOutput[1]);
      expect(errors.length).toBe(0);
      expect(output).toMatchSnapshot();
    });
  });

  describe(Analyzer.prototype.typegen, () => {
    it('should report error when no schema', async () => {
      const analyzer = createTestingAnalyzer(noSchemaPrj);
      const { errors } = await analyzer.typegen();
      expect(errors.length).toBe(1);
      expect(errors[0].message).toMatchSnapshot();
    });

    it('should create type files', async () => {
      const analyzer = createTestingAnalyzer(simpleSources);
      const { outputSourceFiles } = await analyzer.typegen();
      if (!outputSourceFiles) return fail();
      expect(outputSourceFiles.length).toBe(1);
      expect(outputSourceFiles[0].fileName.endsWith('__generated__/my-query.ts')).toBeTruthy();
      expect(outputSourceFiles[0].content).toMatchSnapshot();
    });
  });
});
