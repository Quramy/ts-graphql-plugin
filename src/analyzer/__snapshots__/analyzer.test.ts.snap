// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Analyzer extractToManifest should extract manifest 1`] = `
Array [
  Array [],
  Object {
    "documents": Array [
      Object {
        "body": "query MyQuery {
  hello
}
",
        "documentEnd": Object {
          "character": 41,
          "line": 0,
        },
        "documentStart": Object {
          "character": 18,
          "line": 0,
        },
        "fileName": "main.ts",
        "fragmentName": undefined,
        "operationName": "MyQuery",
        "tag": "gql",
        "templateLiteralNodeEnd": Object {
          "character": 42,
          "line": 0,
        },
        "templateLiteralNodeStart": Object {
          "character": 17,
          "line": 0,
        },
        "type": "query",
      },
    ],
  },
]
`;

exports[`Analyzer report should create markdown report 1`] = `
"# Extracted GraphQL Operations
## Queries

### MyQuery

\`\`\`graphql
query MyQuery {
  hello
}
\`\`\`

From [main.ts:1:19](main.ts#L1-L1)
    
---
Extracted by [ts-graphql-plugin](https://github.com/Quramy/ts-graphql-plugin)"
`;

exports[`Analyzer report should create markdown report from manifest 1`] = `
"# Extracted GraphQL Operations
## Queries

### MyQuery

\`\`\`graphql
query MyQuery {
  hello
}
\`\`\`

From [main.ts:1:19](main.ts#L1-L1)
    
---
Extracted by [ts-graphql-plugin](https://github.com/Quramy/ts-graphql-plugin)"
`;

exports[`Analyzer typegen should create type files 1`] = `
"/* eslint-disable */
/* This is an autogenerated file. Do not edit this file directly! */
export type MyQuery = {
    hello: string;
};
export type MyQueryVariables = {};
"
`;

exports[`Analyzer typegen should report error when no schema 1`] = `"No GraphQL schema. Confirm your ts-graphql-plugin's \\"schema\\" configuration at tsconfig.json's compilerOptions.plugins section."`;

exports[`Analyzer validate should report error when no schema 1`] = `"No GraphQL schema. Confirm your ts-graphql-plugin's \\"schema\\" configuration at tsconfig.json's compilerOptions.plugins section."`;

exports[`Analyzer validate should validate project with schema error project 1`] = `"Syntax Error: Unexpected Name \\"hogehoge\\"."`;
